admin/acceptance.py

```diff
@@ -896,9 +898,41 @@ def capture_journal(reactor, host, output_file):
             b'-u', b'flocker-container-agent',
             b'-u', b'flocker-docker-plugin',
         ],
-        handle_stdout=lambda line: output_file.write(line + b'\n')
+        handle_stdout=formatter,
     )
     ran.addErrback(write_failure, logger=None)
+    # Deliver a final empty line to process the last message
+    ran.addCallback(lambda ignored: formatter(b""))
+
+
+def journald_json_formatter(output_file):
+    """
+    Create an output handler which turns journald's export format back into
+    Eliot JSON with extra fields to identify the log origin.
+    """
+    accumulated = {}
+    def handle_output_line(line):
+        if line:
+            key, value = line.split(b"=", 1)
+            accumulated[key] = value
+        else:
+            if accumulated:
+                raw_message = accumulated.get(b"MESSAGE", b"{}")
+                try:
+                    message = json.loads(raw_message)
+                except ValueError:
+                    # Docker log messages are not JSON
+                    message = dict(message=raw_message)
+
+                message[u"_HOSTNAME"] = accumulated.get(
+                    b"_HOSTNAME", b"<no hostname>"
+                )
+                message[u"_SYSTEMD_UNIT"] = accumulated.get(
+                    b"_SYSTEMD_UNIT", b"<no unit>"
+                )
+                output_file.write(json.dumps(message) + b"\n")
+                accumulated.clear()
```

Comment:
Agree that'd be better factoring.  I'll leave a note but I don't think putting this off 'til later will hurt anything.